lib/ddtrace/sampling/rule_sampler.rb

```diff
@@ -0,0 +1,77 @@
+require 'forwardable'
+
+require 'ddtrace/ext/priority'
+
+require 'ddtrace/ext/sampling'
+require 'ddtrace/sampler'
+require 'ddtrace/sampling/token_bucket'
+
+module Datadog
+  module Sampling
+    # TODO: Write class documentation
+    # RuleSampler
+    class RuleSampler
+      extend Forwardable
+
+      attr_reader :rules, :rate_limiter, :fallback_sampler
+
+      def initialize(rules = [],
+                     rate_limiter = Datadog::Sampling::UnlimitedLimiter.new,
+                     fallback_sampler = Datadog::AllSampler.new)
```

Comment:
This isn't really a "fallback" as it is a "default"

So, we should probably just define this as `default_sample_rate = 1.0`

Then we do `@default_sampler = RateSampler. new(default_sample_rate)` instead of `@fallback_sampler = ...`

Reply:
I think receiving a sampler implementation as the default fallback works better with how we structured our samplers.

The same functionality we can get from having a `default_sample_rate` can be achieved by receiving the sampler instance instead.

I do like the term `default` more than `fallback` though, so I'm renaming the parameter with the suggestion.

Reply:
I think it might actually cause some configuration fatigue if we allow them to pass in any sampler.

Mostly because 1) they shouldn't and 2) we want them to be able to set the default sample rate if no rules match (very common case), if they have to change from an `AllSampler.new` to a `RateSampler.new(rate)`, then it is more that they need to know about vs `DD_DEFAULT_SAMPLE_RATE=0.5` for example.

Reply:
The part where internal vs external APIs diverge in my opinion.

I think the internal API for the `RuleSampler` matches our inner-workings at the correct level.
But I do agree that for external facing configuration it's just too cumbersome.

Reply:
Addressed this. We now have a more humane API :)
```ruby
RuleSampler.new(rules, # Array of rules here
  rate_limit: 1000,
  default_sample_rate: 1.0)
```